models/image.py

```diff
@@ -91,98 +94,101 @@ def format(self, value):
-        )
-    )
+            # make sure the upstream_images list is sorted by mjd:
+            self.upstream_images.sort(key=lambda x: x.mjd)
+            self.ref_image_index = self.upstream_images.index(value)
 
-    new_image = orm.relationship(
-        'Image',
-        cascade='save-update, merge, refresh-expire, expunge',
-        primaryjoin='images.c.new_image_id == images.c.id',
-        uselist=False,
-        remote_side='Image.id',
+    new_image_index = sa.Column(
+        sa.Integer,
+        nullable=True,
         doc=(
-            "New image used to produce a difference image. "
-            "Only set for difference images. This usually refers to a regular image. "
+            "Index of the new image used to produce a difference image, in the upstream_images list. "
         )
     )
 
     @property
-    def is_coadd(self):
-        try:
-            if self.source_images is not None and len(self.source_images) > 0:
-                return True
-        except DetachedInstanceError:
-            if not self.is_sub and self.exposure_id is None:
-                return True
+    def new_image(self):
+        if self.new_image_index is None:
+            return None
+        if len(self.upstream_images) <= self.new_image_index:
+            raise RuntimeError(f'Index {self.new_image_index} is out of range for upstream images!')
+        return self.upstream_images[self.new_image_index]
+
+    @new_image.setter
+    def new_image(self, value):
+        if value is None:
+            self.new_image_index = None
+        else:
+            if not isinstance(value, Image):
+                raise ValueError(f"new_image must be an Image object. Got {type(value)} instead.")
+            if value not in self.upstream_images:
+                raise ValueError(f"new_image must be in the upstream_images list. Got {value} instead.")
 
-        return False
+            # make sure the upstream_images list is sorted by mjd:
+            self.upstream_images.sort(key=lambda x: x.mjd)
+            self.new_image_index = self.upstream_images.index(value)
 
-    @hybrid_property
-    def is_sub(self):
-        try:
-            if self.ref_image is not None and self.new_image is not None:
-                return True
-        except DetachedInstanceError:
-            if self.ref_image_id is not None and self.new_image_id is not None:
-                return True
-
-        return False
+    is_sub = sa.Column(
+        sa.Boolean,
+        nullable=False,
+        default=False,
+        index=True,
+        doc='Is this a subtraction image.'
+    )
 
-    @is_sub.inplace.expression
-    @classmethod
-    def is_sub(cls):
-        return cls.ref_image_id.isnot(None) & cls.new_image_id.isnot(None)
+    is_coadd = sa.Column(
+        sa.Boolean,
+        nullable=False,
+        default=False,
+        index=True,
+        doc='Is this image made by stacking multiple images.'
+    )
```

Comment:
It may well be that sub and coadd are the only compsite images we'll deal with, but if we're going to store flatfields, and flatfied inputs, as images, then there are others.

I wonder, though, if instead of a pair of booleans, it might be worth having a 8-bit or 16-bit integer here that's a "compsite type", where 0 means not a composite.

sub and coadd are two composites.

But, a combined flatfield is a different sort of composite from sub and coadd, as is a combined bias.  (Whether a combined bias and a combined flatfield are the same kind of composite is a longer combination.)  I can't think right now of any other sorts of composites, but we might come up with some in the future, and if we have a byte here for composite type rather than a pair of flags, it's more future proof.

Reply:
I am generally against future proofing.

Reply:
I am also inclined to say we should stop thinking about calibration objects (such as flats) as Image objects. They have much less in common to our definition of Image and I don't think we gain much from asking Image to also represent calibration data. 
I think flats should go as CalibratorFile that are FileOnDiskMixin that happen to be a 2D dataset. But this is a discussion for another time (see Issue #162 )

Reply:
Yeah, I'd agree that we shouldn't think about flatfields as Images.